### ISSUE TITLE
Wrong SQL created by hibernate when using Pageable (creates "limit ?, ?" instead of "limit ? offset ?")

### ISSUE THREAD

**Issue by eugenekup** [olavloite/spanner-hibernate#2]
I am using Spring's `@Query` annotation to write annotation and PagingAndSortingRepository like that: 

`@Query("SELECT ...")
Page<Projection> findProjection(Pageable pageable);`

When Pageable is provided with pageSize > 0, the hibernate creates a query with:
`limit ?, ?` 

However, Spanner expects it to be:
`limit ? offset ?`

So I receive the following error: 

`nl.topicus.jdbc.shaded.io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Syntax error: Unexpected "," [at 1:1455]
... limit @p1, @p2`

I use Cloud SQL and Cloud Spanner databases, my Spanner related config file got the following properties: 

` properties.put("hibernate.schema_management_tool", "nl.topicus.hibernate.dialect.CloudSpannerSchemaManagementTool");
properties.put("hibernate.dialect", "nl.topicus.hibernate.dialect.CloudSpannerDialect");`

I expected that CloudSpannerDialect would 'suggest' hibernate the correct syntax for cloud Spanner re limit and offset. Isn't this the case? 

I use the following maven dependencies: 

<artifactId>spanner-jdbc</artifactId>
<version>1.0.13</version>

<artifactId>spanner-hibernate</artifactId>
<version>0.7</version>

**Comment by olavloite** [comment]
@eugenekup Are you sure your Hibernate dialect setting is being picked up? I don't know what framework(s) you are using, so I don't know if the way you are supplying the properties is correct. The Hibernate dialect should translate these queries into what you suggest (i.e. `select ... limit ? offset ?`).

I have just updated my sample application https://github.com/olavloite/spanner-jpa-example with a similar query as what you suggest above, and in my case the following query is generated:
`select customer0_.id as id1_0_, customer0_.created as created2_0_, customer0_.updated as updated3_0_, customer0_.uuid as uuid4_0_, customer0_.firstName as firstNam5_0_, customer0_.lastName as lastName6_0_ from Customer customer0_ limit ? offset ?`

Could you have a look at this example, or otherwise share a small example project or test case that shows why it isn't working for you?

**Comment by olavloite** [comment]
Oh, and may I suggest that you set up a Cloud Spanner emulator that you could use for these testing purposes. Creating a test instance and test tables is a lot faster on the emulator than on a real Spanner instance.
Have a look here for a tutorial on how to set up a trial account: http://www.googlecloudspanner.com/2018/07/create-trial-account-for-cloud-spanner.html

**Comment by eugenekup** [comment]
@olavloite thanks a lot for your quick response! 

Indeed, when I only use one datasource (all Cloud Spanner related properties specified in application.properties file) the SQL is getting created correctly.

It picks up the correct dialect on the start of the App: 

`org.hibernate.dialect.Dialect            : HHH000400: Using dialect: nl.topicus.hibernate.dialect.CloudSpannerDialect`

So the issue is perhaps with my config. 
It picks up the MySQL55Dialect twice: 

`
com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
org.hibernate.dialect.Dialect            : HHH000400: Using dialect: org.hibernate.dialect.MySQL55Dialect

com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Start completed.
org.hibernate.dialect.Dialect            : HHH000400: Using dialect: org.hibernate.dialect.MySQL55Dialect
`

**Comment by eugenekup** [comment]
Though when I changed the `@Primary` config to be at SpannerConfig class, it started to apply Spanner Dialect. However, the App doesn't start because of this error: 

`Caused by: org.hibernate.QueryException: No data type for node: org.hibernate.hql.internal.ast.tree.MethodNode 
 \-[METHOD_CALL] MethodNode: '('
    +-[METHOD_NAME] IdentNode: 'CONCAT' {originalText=CONCAT}
    \-[EXPR_LIST] SqlNode: 'exprList'
       +-[QUOTED_STRING] LiteralNode: ''RI-PEST-''
       \-[DOT] DotNode: 'employee0_.EmployeeCode' {propertyName=employeeCode,dereferenceType=PRIMITIVE,getPropertyPath=employeeCode,path=e.employeeCode,tableAlias=employee0_,className=com.api.csi.entity.spanner.Employee,classAlias=e}
          +-[ALIAS_REF] IdentNode: 'employee0_.HashId' {alias=e, className=com.api.csi.entity.spanner.Employee, tableAlias=employee0_}
          \-[IDENT] IdentNode: 'employeeCode' {originalText=employeeCode}`

**Comment by olavloite** [comment]
@eugenekup No problem, thanks for letting me know, and don't hesitate to file another issue if you run into something else.

**Comment by eugenekup** [comment]
Hi @olavloite 

Do not want to raise another issue straight away in case it is again some config issue. 

Would you mind confirming that Spanner's `CONCAT(value1[, ...])` function works fine? [Spanner Doc](https://cloud.google.com/spanner/docs/functions-and-operators#concat)

For me, query like that works fine when using `MySQL55Dialect` dialect: 

```
@Query("SELECT e.employeeCode AS employeeCode, CONCAT('CODE-', e.employeeCode) AS fmEmployeeCode FROM Employee e")
  List<TechnicianProjection> findTechnicianDetails();
```

However as soon as I switch to `CloudSpannerDialect` the application fails to start with the error that I also posted in my earlier comment: 

```
Caused by: org.hibernate.QueryException: No data type for node: org.hibernate.hql.internal.ast.tree.MethodNode 
 \-[METHOD_CALL] MethodNode: '('
    +-[METHOD_NAME] IdentNode: '**CONCAT**' {originalText=CONCAT}
    \-[EXPR_LIST] SqlNode: 'exprList'
       +-[QUOTED_STRING] LiteralNode: ''CODE-''
       \-[DOT] DotNode: 'employee0_.EmployeeCode' {propertyName=employeeCode,dereferenceType=PRIMITIVE,getPropertyPath=employeeCode,path=e.employeeCode,tableAlias=employee0_,className=com.api.csi.entity.spanner.Employee,classAlias=e}
          +-[ALIAS_REF] IdentNode: 'employee0_.HashId' {alias=e, className=com.api.csi.entity.spanner.Employee, tableAlias=employee0_}
          \-[IDENT] IdentNode: 'employeeCode' {originalText=employeeCode}
```

This is by no means a show-stopper but I thought to check whether this supposed to work.

**Comment by olavloite** [comment]
@eugenekup I can confirm that the above is a problem in the Cloud Spanner Hibernate dialect. I'll look into it.

**Comment by olavloite** [comment]
@eugenekup There is however a very simple workaround for this specific case. You could define a property on your entity and use a `@Formula` annotation instead of a normal `@Column` annotation. That way, you can easily create the formatted employee code you are looking for like this:

```Java
@Formula("CONCAT('CODE-', employeeCode)")
private String fmEmployeeCode;
```
I have just tried it out with one of my sample applications, and this way it works.

**Comment by eugenekup** [comment]
@olavloite thank you for your advice, however this doesn't work when I use a projection interface.

Entity: 
```java
import org.hibernate.annotations.Formula;

@Getter
@Entity
@Table(name = "Employee")
public class Employee {

  @Column(name = "EmployeeCode")
  private String employeeCode;

  @Formula("CONCAT('CODE-', employeeCode)")
  private String fmEmployeeCode;
}
```

Projection interface: 
```java
public interface TechnicianProjection {
  
  String getEmployeeCode();

  String getFmEmployeeCode();
}
```

Repo:
```java
public interface TechnicianRepo extends CrudRepository<Employee, String> {
  @Query("SELECT e.employeeCode AS employeeCode FROM Employee e")
  List<TechnicianProjection> findTechnicianDetails();
}
```

Returned result: 
```json
{
            "employeeCode": "2929",
            "fmEmployeeCode": null
}
```

I tried both
`@Formula("CONCAT('CODE-', employeeCode)")`
and
`@Formula("CONCAT('CODE-', EmployeeCode)")`

It works fine when I return actual class though. So thank you anyway for this advice, it is very useful.

**Comment by olavloite** [comment]
The problem is that you are not selecting any value for the fmEmployeeCode property. You can select the formula property as if it was a normal property.

If you change your query from

```Java
@Query("SELECT e.employeeCode AS employeeCode FROM Employee e")
```
into

```Java
@Query("SELECT e.employeeCode AS employeeCode, e.fmEmployeeCode as fmEmployeeCode FROM Employee e")
```

 it should work. (The formula should still be `@Formula("CONCAT('CODE-', employeeCode)")`

I have just updated the spanner-jpa-example and added an example using a formula and a projection, and it does work. Have a look at the query on line 30 of https://github.com/olavloite/spanner-jpa-example/blob/master/src/main/java/nl/topicus/spanner/jpa/entities/CustomerRepository.java.

**Comment by eugenekup** [comment]
Thanks a lot @olavloite! This does work very well indeed :) Thank you again for your support (even when it is not your library related issue) 👍

**Comment by olavloite** [comment]
@eugenekup I released version 0.8 of the library last weekend, and that version should fix the problem you had with the CONCAT function.
